### DISCUSSION TITLE
Writing an interstitial filed to the history file

### DISCUSSION THREAD

**Discussion by ericaligo-NOAA** [NOAA-EMC/ufsatm#486] [Q&A]
I would like to write out graupelmp, an interstitial, to the FV3 history file, however, it's not clear if and how this can be added to GFS_diagnostics.F90.  An attempt has been made to add this field to GFS_diagnostics.F90, but a compiler error notes that interstitial and IntDiag can not be used. What options do I have?
Eric

**Comment by climbfuji** [comment]
You can't output interstitial data to disk. The data is not persistent and does not contain valid data at the time the diagnostic output is written.

If you just want to output the existing array as it is, you could move it from the Interstitial DDT to a persistent DDT (e.g. Sfcprop). All it takes is moving a bit of Fortran code in `GFS_typedefs.F90`, moving the metadata in `GFS_typedefs.meta`, and then add the field in `GFS_diagnostics.F90`.

**Reply by ericaligo-NOAA** [reply]
I moved it to Diag instead in all of the places that I'm aware of, however, I was getting the error message below.  I'm not sure why it still thinks rainmp is an interstitial.  Is this okay for me to move rainmp to Diag instead of Sfcprop?
/scratch2/NCEPDEV/fv3-cam/save/Eric.Aligo/ufs-weather-model_winterwx/build/FV3/ccpp/physics/ccpp_FV3_GFS_v15_thompson_mynn_lam3km_physics_cap.F90(2077): error #6460: This is not a field name that is defined in the encompassing structure.   [RAINMP]
                  rain=GFS_Interstitial(cdata%thrd_no)%rainmp,graupel=GFS_Interstitial(cdata%thrd_no)%graupelmp, &
-------------------------------------------------------^

I also modified GFS_debug.F90 and moved rainnmp from interstitial to Diag:
                     call print_var(mpirank, omprank, blkno, Grid%xlat_d, Grid%xlon_d, 'Diag%rainmp      ',    Diag%rainmp)

**Comment by climbfuji** [comment]
If the framework thinks it is in the interstitial DDT, then you need to move it in GFS_typedefs.F90 from within the interstitial metadata table into the diag metadata table.

diag is ok but you want to make sure that those quantities get reset to zero each time or are intent(out) variables so that there is no accumulation.

**Reply by ericaligo-NOAA** [reply]
Ohhhhhh wow!  I missed that one.  Thanks.  Well, I 'do' want these accumulated. I followed what was done for totprcp and 'not' totprcpb, which is a bucket precip field.  See below where I moved rainmp,graupelmp and snowmp to.  

I didn't change intent anywhere though. I could easily move it Sfcprop if that means I don't need to worry about intent in all of the routines, but I can still accumulate. What do you suggest?

    if (present(linit) ) set_totprcp = linit
    if (present(iauwindow_center) ) set_totprcp = iauwindow_center
    if (set_totprcp) then
      Diag%totprcp = zero
      Diag%cnvprcp = zero
      Diag%totice  = zero
      Diag%totsnw  = zero
      Diag%totgrp  = zero
      **Diag%graupelmp  = zero
      Diag%rainmp     = zero
      Diag%snowmp     = zero**

**Comment by climbfuji** [comment]
I think you need to check inside mp_thompson.F90 if the logic overwrites the fields or adds to it, apart from that your logic for diag reset seems to be ok.

**Reply by ericaligo-NOAA** [reply]
Exactly.  Graupelmp, rainmp and snowmp in GFS_typdefs.F90, also referred to as graupel, rain and snow in mp_thompson.F90, are accumulated in mp_thompson.F90:

         !> - Convert rainfall deltas from mm to m (on physics timestep); add to inout variables
         ! "rain" in Thompson MP refers to precipitation (total of liquid rainfall+snow+graupel+ice)
         prcp    = prcp    + max(0.0, delta_rain_mp/1000.0_kind_phys)
         graupel = graupel + max(0.0, delta_graupel_mp/1000.0_kind_phys)
         ice     = ice     + max(0.0, delta_ice_mp/1000.0_kind_phys)
         snow    = snow    + max(0.0, delta_snow_mp/1000.0_kind_phys)
         rain    = rain    + max(0.0, (delta_rain_mp - (delta_graupel_mp + delta_ice_mp + delta_snow_mp))/1000.0_kind_phys)

I often search the standard name because that seems to connect what I see in the /physics meta files to what I see in GFS_typedefs.meta:

mp_thompson.meta:
[rain]
  **standard_name = lwe_thickness_of_explicit_rain_amount**
  long_name = explicit rain fall on physics timestep
  units = m
  dimensions = (horizontal_loop_extent)
  type = real
  kind = kind_phys
  intent = inout

GFS_typedefs.F90
[rainmp]
  standard_name = lwe_thickness_of_explicit_rain_amount
  long_name = explicit rain on physics timestep
  units = m
  dimensions = (horizontal_loop_extent)
  type = real
  kind = kind_phys

I removed the line below from the GFS_typedefs.meta file for rainmp,snowmp and graupelmp because I removed it from this if condition in GFS_typedefs.F90:
 active = (control_for_microphysics_scheme == identifier_for_gfdl_microphysics_scheme .or. control_for_microphysics_scheme == identifier_for_thompson_microphysics_scheme)

**Reply by ericaligo-NOAA** [reply]
Hmm, I guess I could keep the if condition in GFS_typdefs.F90, but I'd have to move it out of the interstitial section and put it in the Diag section.  Below shows that I commented out the allocation for those 3 arrays within the if condition.

   ! Allocate arrays that are conditional on physics choices
    if (Model%imp_physics == Model%imp_physics_gfdl .or. Model%imp_physics == Model%imp_physics_thompson) then
!       allocate (Interstitial%graupelmp  (IM))
       allocate (Interstitial%icemp      (IM))
!       allocate (Interstitial%rainmp     (IM))
!       allocate (Interstitial%snowmp     (IM))

I would then add the active line back to those 3 fields in the meta file:
active = (control_for_microphysics_scheme == identifier_for_gfdl_microphysics_scheme .or. control_for_microphysics_scheme == identifier_for_thompson_microphysics_scheme)
